api/agent/agent.go

```diff
@@ -748,19 +746,25 @@ type container struct {
 	stats *drivers.Stats
 }
 
-func (c *container) swap(stderr io.Writer, cs *drivers.Stats) (old io.Writer) {
+func (c *container) swap(stdin io.Reader, stdout, stderr io.Writer, cs *drivers.Stats) func() {
 	c.Lock()
 	defer c.Unlock()
 
-	// TODO meh, maybe shouldn't bury this
-	old = c.stderr
-	gw, ok := c.stderr.(*ghostWriter)
-	if ok {
-		old = gw.swap(stderr)
-	}
+	ostdin := c.stdin.(*ghostReader).inner
+	ostdout := c.stdout.(*ghostWriter).inner
+	ostderr := c.stderr.(*ghostWriter).inner
 
+	// if tests don't catch this, then fuck me
+	c.stdin.(*ghostReader).swap(stdin)
+	c.stdout.(*ghostWriter).swap(stdout)
+	c.stderr.(*ghostWriter).swap(stderr)
 	c.stats = cs
-	return old
+
+	return func() {
+		c.stdin.(*ghostReader).swap(ostdin)
+		c.stdout.(*ghostWriter).swap(ostdout)
+		c.stderr.(*ghostWriter).swap(ostderr)
```

Comment:
couple questions:
1) how about stats? shouldn't we set stats to nil here to stop accumulating stats via ?
2) shouldn't we lock/unlock in the function returned?
3) unrelated but 'lock' as inherited sync.Mutex in container struct makes this harder to understand IMHO, what does that lock protect? entire container? or just stats and io? We pass the container to Docker driver and that reads container fields without locking it... which means, we only protect IO and stats swap/append with it, right? (I'll fix/rename this later.)

Reply:
I think the lock here is only needed for stats for driver collect go routine (append) and during swap (set/unset) of stats member variable. There's no other go routine trying to swap these at the same time.

Reply:
Aren't we allocating new slice for container stats for each call? As far as understand, we do that each time we call `task.WriteStat`

Reply:
is calling append() with nil safe? There's a nil check there... in WriteStat()

But yeah, I think we only need to protect stats here. Lock the mutex when setting/unsetting/append...

Reply:
In other words, if stats is nil, we shouldn't collect stats, right? That's what WriteStat() is trying to do I think.

Reply:
thanks, good catch. yes, we need to set stats to nil and have the mutex to swap just for stats I think. the ghosts have their own locks. 

> unrelated but 'lock' as inherited sync.Mutex in container struct makes this harder to understand IMHO, what does that lock protect? entire container? or just stats and io? 

think it's expanded and contracted a few times: id, stats, writers. now the reader and writers have their own locks. will make this more clear, thanks, it is definitely not a 'container' lock as implied currently.

Reply:
we can merge, and then tweak this as a followup as this should work fine.

Reply:
gonna fix the stats mu shortly, then i think we're safe